keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+resources after they no longer need those resources. For example, a Job gets
+deleted 1 day after it completes, a Secret gets deleted 30 minutes after it is
+no longer used by another other resources, and a Pod without an owner gets
+deleted 1 hour after it terminates.
+
+[mark-and-sweep]: https://www.geeksforgeeks.org/mark-and-sweep-garbage-collection-algorithm/
+
+### Summary
+
+In summary, resource garbage collection with TTL is illustrated below:
+
+![Garbage Collection Diagram](garbage-collection-diagram.png)
+
+The orange boxes in the above diagram represents the new design of the garbage
+collector. 
+
+When all garbage collection mechanisms agree a resource should be deleted, the
+resource becomes "garbage collection candidate" and then the user-specified TTL
+kicks in. If somehow the resource leaves the "garbage collection candidate"
+state, the TTL will be reset. The resource will only be deleted when it is a
+garbage collection candidate and its TTL has expired (zero TTL expires
+immediately). Garbage collector will implement its own TTL mechanism instead of
+using etcd TTL.
+
+Note that current Pod garbage collection ([capacity-based][]) and Event garbage
+collection ([TTL-based][], using etcd TTL) are not handled by the garbage
+collector. They are kept as-is for backward compatibility. In the future, we may
+move them into garbage collector, or deprecate them and delete them entirely.
+
+[capacity-based]: #non-generic-capacity-based-garbage-collection
+[TTL-based]: #non-generic-ttl-based-garbage-collection
+
+### User Stories
+
+#### Story 1: Clean up old Jobs
+
+The users keep creating Jobs in a small Kubernetes cluster with 4 nodes. The
+Jobs accumulates over time, and 1 year later, the cluster ended up with more
+than 100k old Jobs. This caused etcd hiccups, long high latency etcd requests,
+and eventually made the cluster unavailable.
+
+The problem could have been avoided easily by enabling TTL-based garbage
+collection with Jobs completion time. 
+
+The steps are as easy as:
+
+1. In Job’s metadata, the user sets TTL to 1 hour and that the Job should be
+   cleaned up after it completes (unused).
+1. The user deploys Jobs as usual.
+1. After the Job completes, the result is observed asynchronously within an hour
+   and stored elsewhere. 
+1. The garbage collector cleans up those Jobs 1 hour after Job completion. 
+
+#### Story 2: Secrets and ConfigMaps
+
+Secrets and ConfigMaps are used and referenced by other resources, such as Pods,
+ReplicaSets, and Deployments. An advised pattern is to treat application
+configurations as immutable so that application revisions are intact and
+rollbacks are reliable. When the users deploy their applications, they treat
+Secrets and ConfigMaps as immutable resources, and therefore Secrets and
+ConfigMaps are frequently created and deployed to a Kubernetes cluster, but
+never updated or cleaned up. Secrets and ConfigMaps accumulate over time, which
+could cause master component downtime. 
```

Comment:
+1 for having immutable Secrets, but I am neutral on ConfigMaps. I think ConfigMaps should provide enough versatility for minor config or script injection changes.

Reply:
The downside of updating ConfigMaps is that it doesn't work well with rolling update features for workloads APIs. You don't have control over when a ConfigMap change will be updated to a Pod, and you cannot roll back.

Reply:
We were/are actually thinking about creating a top level object to allow for ConfigMap/Secret snapshoting and also triggering rollouts. (It would also have a revisionHistoryLimit and be bound to particular RS or ControllerRevision.)    /cc @mfojtik 

I don't think the Secret/ConfigMap should be immutable per type, but rather per instance - like having metadata.immutable=true as there are always different use cases across the ecosystem.

Reply:
+1 on @tnozicka 's suggestion on creating a top level object for managing ConfigMap/Secret snapshots and rollouts.

If we are going to introduce a field indicating immutability preference, it may be better to add the field to a standalone `.immutability` field outside of `.metadata`.